### PR TITLE
[WIP] High-level Architecture

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Just starting some ideas / stealing them from other places.
Gotta figure out what all needs to be implemented.

Goals from @beauby : 
- strong typing
- handles the data locally (updating, creating, etc.)
- handles querying the remote API
- handles inverse relationships upon update (i.e. when fetching a collection of posts with related comments, if some comments that were locally linked to a post are no longer, remove the local inverse relationship, stuff like that)

### CODE REVIEW

File: Architecture.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Architecture
+
+This is just a proposal of a potential set of classes / services and their responsibilities and what APIs they may have.
+
+## Top-Level Objects
+
+<table>
```
Comment (NullVoxPopuli) [inline Architecture.md:7]: sorry for using HTML ;-)

File: Architecture.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  <thead>
+    <tr><th style='width: 120px'>Object</th><th>Description / Responsibility</th><th>Example APIs</th>
+  </thead>
+  <tbody>
+    <tr>
+      <td>Model</td><td>defines the attributes and relationships for resources </td>
+      <td>
+        - save()<br>
```
Comment (beauby) [inline Architecture.md:15]: So in order to have those methods on the model, we need to have the model access the store (via dependency injection).

File: Architecture.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        - performs CRUD on any resource
+      </td>
+      <td>
+        - createRecord(type, params)<br>
+        - findRecord(type, options)<br>
+        - peekRecord(type, id)<br>
+        - findAll(type)<br>
+        - peekAll(type)<br>
+        - query(type, params)<br>
+        - queryRecord(type, params)<br>
```
Comment (beauby) [inline Architecture.md:38]: We might want to add a method to sync an arbitrary payload into the store (for testing for instance).

File: Architecture.md (hunk 2 above)
Comment (NullVoxPopuli) [inline Architecture.md:15]: yeah, possibly. I was thinking the models could be a subclass of a `NGModel` defined in this project that would include a reference to the store or something.

File: Architecture.md (hunk 3 above)
Comment (NullVoxPopuli) [inline Architecture.md:38]: yeah, EmberData calls that `pushPayload`. Definitely something to add.

**Comment by beauby** [comment]
I'm wondering whether we should expose a service for each model, so that models are just that, models.

**Comment by NullVoxPopuli** [comment]
Cleaning up my PRs. If this work is still desired, feel free to cherry-pick

**Pull request closed without merging**